include/maths/CSampling.h

```diff
@@ -42,6 +45,71 @@ class MATHS_EXPORT CSampling : private core::CNonInstantiatable {
     using TSizeVec = std::vector<std::size_t>;
     using TPtrdiffVec = std::vector<std::ptrdiff_t>;
 
+    //! \brief A basic stream sampler.
+    template<typename T>
+    class CStreamSampler {
+    public:
+        virtual ~CStreamSampler() = default;
+
+        //! Get the sampler's random number generator.
+        CPRNG::CXorOShiro128Plus& rng() { return m_Rng; }
+
+        //! Get the sample size.
+        std::size_t sampleSize() const { return m_SampleSize; }
+
+        //! Get the size of the sample we're targeting.
+        std::size_t targetSampleSize() const { return m_TargetSampleSize; }
+
+        //! Reset in preparation for resampling.
+        virtual void reset() {
+            m_SampleSize = 0;
+            m_StreamWeight = m_SampleWeight = 0.0;
+        }
+
+        //! Sample the value \p x uniformly at random.
+        void sample(const T& x) {
```

Comment:
Can you please explain the sampling algorithm or give a reference to the description?

Reply:
I'll write some fuller comments.

Ideally I'd use a full quasi-random sampling scheme, but I wanted something v.fast and streaming. In the end, I adapted the idea behind k-means++ initialisation, but wanted to avoid having to do nearest neighbour search to get the probability with which to sample each point. This was also based on trying to get something in place quickly given closeness of 7.7 (and we maybe can refine).

So what does this gain us?

For interest, here is an example of some testing I did using random logit values for the leaf predictions. (We get similar behaviour for a variety of distributions and often noticeably lower errors by spreading out the samples better.)

<img width="526" alt="Screenshot 2020-03-25 at 18 30 10" src="https://user-images.githubusercontent.com/7591487/77572864-a4c2d280-6ec7-11ea-94d6-5000978cfc6e.png">

The error in approximate loss surface using random samples:
<img width="509" alt="Screenshot 2020-03-25 at 18 29 17" src="https://user-images.githubusercontent.com/7591487/77572904-b2785800-6ec7-11ea-9478-01b63d7141a8.png">

The error in the approximate loss surface using dissimilar samples:
<img width="500" alt="Screenshot 2020-03-25 at 18 30 42" src="https://user-images.githubusercontent.com/7591487/77572999-e784aa80-6ec7-11ea-8cf3-e9b1bc278f82.png">

Flattening the error surfaces:
<img width="580" alt="Screenshot 2020-03-25 at 18 41 29" src="https://user-images.githubusercontent.com/7591487/77573235-4518f700-6ec8-11ea-8717-54f27c813d0f.png">

Reply:
I added a fuller description for the new sampling in [34302ba](https://github.com/elastic/ml-cpp/pull/1082/commits/34302ba97a86af44219f12a1129f1286e1168f0d).